File: app/src/main/java/se/creotec/chscardbalance2/model/BackendResponse.java
// Copyright (c) 2017 Alexander Håkansson
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
package se.creotec.chscardbalance2.model;

public class BackendResponse<T> {
    public static final String UNKNOWN_ERROR = "Unknown error";

    private boolean success;
    private String errorMessage;
    private T data;

    public BackendResponse() {
        this.success = false;
        this.errorMessage = UNKNOWN_ERROR;
        this.data = null;
    }

    public boolean isSuccess() {
        return success;
    }

    public void setSuccess(boolean success) {
        this.success = success;
    }

    public String getErrorMessage() {
        return errorMessage;
    }

    public void setErrorMessage(String errorMessage) {
        this.errorMessage = errorMessage;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        BackendResponse<?> that = (BackendResponse<?>) o;

        if (success != that.success) return false;
        if (errorMessage != null ? !errorMessage.equals(that.errorMessage) : that.errorMessage != null)
            return false;
        return data != null ? data.equals(that.data) : that.data == null;

    }

    @Override
    public int hashCode() {
        int result = (success ? 1 : 0);
        result = 31 * result + (errorMessage != null ? errorMessage.hashCode() : 0);
        result = 31 * result + (data != null ? data.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "BackendResponse{" +
                "success=" + success +
                ", errorMessage='" + errorMessage + '\'' +
                ", data=" + data +
                '}';
    }
}

File: app/src/main/java/se/creotec/chscardbalance2/model/BackendResponse.kt
// Copyright (c) 2017 Alexander Håkansson
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
package se.creotec.chscardbalance2.model

import com.google.gson.annotations.SerializedName

class BackendResponse<T> {

    @SerializedName("success")
    var isSuccess: Boolean = false
    @SerializedName("error")
    var errorMessage: String = UNKNOWN_ERROR
    @SerializedName("data")
    var data: T? = null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as BackendResponse<*>
        if (isSuccess != that.isSuccess || errorMessage != that.errorMessage) return false
        return if (data != null) data == that.data else that.data == null
    }

    override fun hashCode(): Int {
        var result = if (isSuccess) 1 else 0
        result = 37 * result + errorMessage.hashCode()
        result = 37 * result + (data?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "BackendResponse{" +
                "success=" + isSuccess +
                ", errorMessage='" + errorMessage + '\'' +
                ", data=" + data +
                '}'
    }

    companion object {
        val UNKNOWN_ERROR = "Unknown error"
    }
}
